cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -15,44 +15,446 @@
+
+ protected:
+  std::shared_ptr<DataType> type_;
+};
+
+template <typename Builder>
+Status MakeBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
+                   std::unique_ptr<Builder>* out) {
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(pool, type, &builder));
+  out->reset(checked_cast<Builder*>(builder.release()));
+  return Status::OK();
 }
 
-Status FilterBinaryKernel::Call(FunctionContext* ctx, const Datum& left,
-                                const Datum& right, Datum* out) {
-  DCHECK(left.type()->Equals(right.type()));
+template <typename Builder, typename Scalar>
+static Status UnsafeAppend(Builder* builder, Scalar&& value) {
+  builder->UnsafeAppend(std::forward<Scalar>(value));
+  return Status::OK();
+}
+
+static Status UnsafeAppend(BinaryBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+static Status UnsafeAppend(StringBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+// TODO(bkietz) this can be optimized
+static int64_t OutputSize(const BooleanArray& filter) {
+  auto offset = filter.offset();
+  auto length = filter.length();
+  internal::BitmapReader filter_data(filter.values()->data(), offset, length);
+  int64_t size = 0;
+  for (auto i = offset; i < offset + length; ++i) {
+    if (filter.IsNull(i) || filter_data.IsSet()) {
+      ++size;
+    }
+    filter_data.Next();
+  }
+  return size;
+}
+
+template <typename ValueType>
+class FilterImpl;
+
+template <>
+class FilterImpl<NullType> : public FilterKernel {
+ public:
+  using FilterKernel::FilterKernel;
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    out->reset(new NullArray(OutputSize(checked_cast<const BooleanArray&>(filter))));
+    return Status::OK();
+  }
+};
+
+template <typename ValueType>
+class FilterImpl : public FilterKernel {
+ public:
+  using ValueArray = typename TypeTraits<ValueType>::ArrayType;
+  using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+
+  using FilterKernel::FilterKernel;
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    std::unique_ptr<OutBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), type_, &builder));
+    RETURN_NOT_OK(builder->Resize(OutputSize(filter)));
+    RETURN_NOT_OK(UnpackValuesNullCount(checked_cast<const ValueArray&>(values), filter,
+                                        builder.get()));
+    return builder->Finish(out);
+  }
+
+ private:
+  Status UnpackValuesNullCount(const ValueArray& values, const BooleanArray& filter,
```

Comment:
@fsaintjacques currently I only use this optimization for primitive types. Do you think it's worthwhile to add it into the nested types as well?